### PR TITLE
auth-layer-proxy improvements for envoy configuration configurable

### LINKED ISSUE #49: Auth-Layer-Proxy: Improvements for configurable parameters for envoy config yaml
Is needed to make the configuration of envoy-auth.yaml config, configurable using `.env` environment variables that can be set from a `config map` on the k8 environment, as well as defined as part of the `values.yaml` when deploying the application using a helm chart.

### PR DESCRIPTION
**Pull request by AlfredoG87**
**Description**:
Changed config of envoy to a template, and added a start-envoy.sh script that is run on initialization that creates the actual config file using the template and the env config variables

changes to the `auth-layer-proxy` to support configuration on deployment, in preparation for the helm charts.
- moved everything inside the docker image to `/etc/envoy/` to keep the standard structure of envoy original image.
- replaced the custom config for a config-template with placeholders for the values that will be replaced from env config.
- created a `start-envoy.sh` script instead of directly starting the envoy with custom config.
- start-envoy.sh takes the config-template and creates a config by applying the environment configurations
- did the necessary changes for supporting the template approach to dockerfile
- did the corresponding changes for the `docker-compose.yml` when running it locally

**Related issue(s)**:
#

Fixes #49

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: auth-layer-proxy/Dockerfile [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
+# Copy all necessary files
+COPY /filters/ /etc/envoy/filters/
+COPY /configs/ /etc/envoy/configs/
+COPY /scripts/start-envoy.sh /etc/envoy/start-envoy.sh
 
-# Install Lua and Luarocks
-RUN     apt-get update && apt-get install -y lua5.1 luarocks git
+# Make the start script executable, change ownership, install dependencies, and clean up in a single RUN to reduce layers
+RUN chmod +x /etc/envoy/start-envoy.sh && \
+    chown -R envoy:envoy /etc/envoy && \
```
Comment (Nana-EC) [inline auth-layer-proxy/Dockerfile:10]: Check the Wiz Inc. recommendations here to see if there are any valuable changes to make

File: auth-layer-proxy/README.md [hunk 2]
```diff
@@ -93,40 +93,58 @@ For instructions on how to set-up the Auth Provider using KeyCloak, refer to the
...
+| `TOKEN_INTROSPECTION_URL` | OAuth Token Introspection URL, provided by the auth server | `http://host.docker.internal:8080/realms/HederaTheGraph/protocol/openid-connect/token/introspect` |
+
 
+### Configure the details of the service to be proxied on the envoy.yaml
+EnvoyProxy needs a configuration file to run, the configuration will be created or updated on the container start, by the `scripts/start-envoy.sh` script, which will create or replace the `envoy-config.yaml` file using the template `configs/envoy-auth-template.yaml` and the environment variables.
+
+It will also print out the configuration as part of the logs for debugging and verification purposes.
+
+Is important to note that if the downstream service that we are protecting, in this case TheGraph, will be accessed by the proxy using a FQDN, the `SERVICE_TYPE` should be set to `LOGICAL_DNS` and the `SERVICE_ADDRESS` should be set to the FQDN of the service, otherwise, if the downstream service is accessed by the proxy using an IP address, the `SERVICE_TYPE` should be set to `STATIC` and the `SERVICE_ADDRESS` should be set to the IP address of the service.
```
Comment (Nana-EC) [inline auth-layer-proxy/README.md:134]: Same suggestion that I made in PR 46.

File: auth-layer-proxy/configs/envoy-auth-template.yaml [hunk 3]
```diff
@@ -32,18 +32,24 @@ static_resources:
...
       - lb_endpoints:
         - endpoint:
             address:
               socket_address:
-                address: host.docker.internal
-                port_value: 8020
+                address: ${SERVICE_ADDRESS}
+                port_value: ${SERVICE_PORT}
+
+admin:
```
Comment (Nana-EC) [inline auth-layer-proxy/configs/envoy-auth-template.yaml:51]: If possible order the yaml sections alphabetically, so `admin` would be up top somewhere

File: auth-layer-proxy/scripts/start-envoy.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+echo "----------------------------------------"
+echo "---  Environment variables:          ---"
+echo "----------------------------------------"
+echo "SERVICE_TYPE: $SERVICE_TYPE"
+echo "SERVICE_ADDRESS: $SERVICE_ADDRESS"
+echo "SERVICE_PORT: $SERVICE_PORT"
+echo "ENVOY_ADMIN_PORT: $ENVOY_ADMIN_PORT"
+echo "PROXY_PORT: $PROXY_PORT"
+echo "CLIENT_ID: $CLIENT_ID"
+echo "CLIENT_SECRET: $CLIENT_SECRET"
```
Comment (Nana-EC) [inline auth-layer-proxy/scripts/start-envoy.sh:22]: Q: should we be echoing this?

**Review by Nana-EC — COMMENTED** [review]
Some suggestions

File: auth-layer-proxy/scripts/start-envoy.sh (hunk 4 above)
Comment (AlfredoG87) [reply auth-layer-proxy/scripts/start-envoy.sh:22]: yes, we should not log the secret, maybe we can obfuscate it and only show the last 4 bytes that could be useful for troubleshooting and make sure that the `auth-layer-proxy` is using the correct configuration values.

File: auth-layer-proxy/README.md (hunk 2 above)
Comment (AlfredoG87) [reply auth-layer-proxy/README.md]: ✅  Done 👍

File: auth-layer-proxy/configs/envoy-auth-template.yaml (hunk 3 above)
Comment (AlfredoG87) [reply auth-layer-proxy/configs/envoy-auth-template.yaml]: ✅  Done! 👍

File: auth-layer-proxy/scripts/start-envoy.sh (hunk 4 above)
Comment (AlfredoG87) [reply auth-layer-proxy/scripts/start-envoy.sh]: decided to go for the first 4 chars:

it looks something like this:

```bash
CLIENT_SECRET: 0cyY******
```

✅  Done! 👍

File: auth-layer-proxy/scripts/start-envoy.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+# Use envsubst to replace environment variables in the template
+envsubst < /etc/envoy/configs/envoy-auth-template.yaml > /etc/envoy/envoy-config.yaml
+
+# Print the processed configuration
+echo "----------------------------------------"
+echo "---  Processed Envoy configuration:  ---"
+echo "----------------------------------------"
+cat /etc/envoy/envoy-config.yaml
```
Comment (Nana-EC) [inline auth-layer-proxy/scripts/start-envoy.sh:36]: Q: Just checking that this doesn't defeat the purpose of the first 4 chars i.e. does this cat expose the secret in logs also?

**Review by Nana-EC — COMMENTED** [review]
LG, 1 clarification

File: auth-layer-proxy/scripts/start-envoy.sh (hunk 5 above)
Comment (AlfredoG87) [reply auth-layer-proxy/scripts/start-envoy.sh:36]: no it does not, the client_secret is used directy on the `TokenVerificationFilter.lua`
https://github.com/hashgraph/hedera-the-graph/blob/main/auth-layer-proxy/filters/TokenVerificationFilter.lua#L25

the output of this is useful for troubleshooting and verifying that the configuration was processed as desired and it looks like this:

```yaml
envoy-1  | ----------------------------------------
envoy-1  | ---  Environment variables:          ---
envoy-1  | ----------------------------------------
envoy-1  | SERVICE_TYPE: LOGICAL_DNS
envoy-1  | SERVICE_ADDRESS: host.docker.internal
envoy-1  | SERVICE_PORT: 8020
envoy-1  | ENVOY_ADMIN_PORT: 16000
envoy-1  | PROXY_PORT: 10000
envoy-1  | CLIENT_ID: htg-auth-layer
envoy-1  | CLIENT_SECRET: 0cyY******
envoy-1  | TOKEN_INTROSPECTION_URL: http://host.docker.internal:8080/realms/HederaTheGraph/protocol/openid-connect/token/introspect
envoy-1  | ----------------------------------------
envoy-1  | ----------------------------------------
envoy-1  | ---  Processed Envoy configuration:  ---
envoy-1  | ----------------------------------------
envoy-1  | admin:
envoy-1  |   address:
envoy-1  |     socket_address:
envoy-1  |       address: 0.0.0.0
envoy-1  |       port_value: 16000
envoy-1  | 
envoy-1  | static_resources:
envoy-1  |   clusters:
envoy-1  |   - name: upstream_service
envoy-1  |     connect_timeout: 5s
envoy-1  |     type: LOGICAL_DNS
envoy-1  |     load_assignment:
envoy-1  |       cluster_name: upstream_service
envoy-1  |       endpoints:
envoy-1  |       - lb_endpoints:
envoy-1  |         - endpoint:
envoy-1  |             address:
envoy-1  |               socket_address:
envoy-1  |                 address: host.docker.internal
envoy-1  |                 port_value: 8020
envoy-1  | 
envoy-1  |   listeners:
envoy-1  |   - name: listener_0
envoy-1  |     address:
envoy-1  |       socket_address:
envoy-1  |         address: 0.0.0.0
envoy-1  |         port_value: 10000
envoy-1  |     filter_chains:
envoy-1  |     - filters:
envoy-1  |       - name: envoy.filters.network.http_connection_manager
envoy-1  |         typed_config:
envoy-1  |           "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
envoy-1  |           stat_prefix: edge
envoy-1  |           http_filters:
envoy-1  |           - name: envoy.filters.http.lua
envoy-1  |             typed_config:
envoy-1  |               "@type": type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua
envoy-1  |               default_source_code: 
envoy-1  |                 filename: /etc/envoy/filters/TokenVerificationFilter.lua
envoy-1  | 
envoy-1  |           - name: envoy.filters.http.router
envoy-1  |             typed_config:
envoy-1  |               "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router          
envoy-1  |           route_config:
envoy-1  |             virtual_hosts:
envoy-1  |             - name: all_domains
envoy-1  |               domains: ["*"]
envoy-1  |               routes:
envoy-1  |               - match:
envoy-1  |                   prefix: "/"
envoy-1  |                   headers:
envoy-1  |                     name: ":method"
envoy-1  |                     exact_match: "POST"
envoy-1  |                 route:
envoy-1  |                   cluster: upstream_service
envoy-1  | ----------------------------------------
```

**Review by Nana-EC — APPROVED** [review]

**Pull request merged**